ios/engine/KMEI/KeymanEngine/Classes/Resource Management/ResourceDownloadQueue.swift

```diff
@@ -9,82 +9,98 @@
 import Foundation
 import Reachability
 
-protocol DownloadNode {
-  var resources: [AnyLanguageResource]? { get }
-}
+enum DownloadNode {
+  case simpleBatch(AnyDownloadBatch)
+  case compositeBatch(CompositeBatch)
 
-class DownloadTask: DownloadNode {
-  public enum Resource {
-    case keyboard, lexicalModel, other
+  var resources: [AnyLanguageResource] {
+    switch(self) {
+      case .simpleBatch(let batch):
+        return batch.resources
+      case .compositeBatch(let node):
+        return node.resources
+    }
   }
-  
-  public final var type: Resource
-  public final var resources: [AnyLanguageResource]?
+
+  var errors: [Error?] {
+    get {
+      switch(self) {
+        case .simpleBatch(let batch):
+          return batch.errors
+        case .compositeBatch(let node):
+          return node.errors
+      }
+    }
+
+    set(value) {
+      switch(self) {
+        case .simpleBatch(var batch):
+          batch.errors = value
+        case .compositeBatch(let node):
+          node.errors = value
+      }
+    }
+  }
+}
+
+protocol AnyDownloadTask {
+  var type: LanguageResourceType? { get }
+  var request: HTTPDownloadRequest { get }
+}
+
+class DownloadTask<Resource: LanguageResource>: AnyDownloadTask {
+  public final var type: LanguageResourceType?
+  public final var resources: [Resource]?
```

Comment:
While we're still performing Cloud-API downloads, we need to keep the actual `Resource` references from the Cloud API search, as that metadata is needed to auto-generate kmp.jsons for cloud resources.

Once we are able to fully convert to pure KMP downloads, this should be reduced to just the `FullID` - that'll be sufficient to install desired resources from KMPs.